### PR TITLE
Publish docker images via GHA

### PR DESCRIPTION
**Pull request by DMRobertson**
I tested this by [running on my repo](https://github.com/DMRobertson/sytest/actions/workflows/docker.yml) with an access token for [my dockerhub account](https://hub.docker.com/u/dmrobertsonelement).

[This run](https://github.com/DMRobertson/sytest/actions/runs/1217188701) shows what a successful workflow should look like.

To publish these images to `matrixdotorg` we'll need an access token secret in this repository. I've requsted that in matrix-org/matrix-ansible-private#3593.

### CODE REVIEW

File: docker/build.sh [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #! /usr/bin/env bash
```
Comment (reivilibre) [inline docker/build.sh:1]: I wonder if we want to remove these (or at least just `push.sh`) if they're obsolete? Having a single way to do things seems sensible/less confusing.

**Review by reivilibre — COMMENTED** [review]
This looks sensible to me (don't forget to remove `dmr/publish-docker-images` from the branch list before you merge ;p), but a couple of questions which probably want feedback from someone more opinionated.

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        uses: docker/setup-buildx-action@v1
+
+      - name: Inspect builder
+        run: docker buildx inspect
+
+      - name: Log in to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_HUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
```
Comment (reivilibre) [inline .github/workflows/docker.yml:38]:
```suggestion
          password: ${{ secrets.DOCKER_HUB_TOKEN }}
```

There's an org-wide secret which is individually granted to specific repos with this name (if I understand properly)

File: docker/README.md [hunk 3]
```diff
@@ -107,8 +107,10 @@ docker run --rm -it ... matrixdotorg/sytest-synapse:buster tests/20profile-event
 
 ## Building the containers
 
-The containers are built by executing `./build.sh`. You will then have to push
-them up to Docker Hub with `./push.sh`.
+These are automatically build and deployed on the `develop` branch by GitHub Actions.
+See `.github/workflows.docker.yml` for the gory details.
+
+Historically, the containers were built by executing `./build.sh`, then pushed up to Docker Hub with `./push.sh`.
```
Comment (reivilibre) [inline docker/README.md:113]: Perhaps we just want to remove these lines? If someone needs to know how things were done in a previous life, git blame can bring them back to the history — it *feels like* it makes sense for these kinds of living documents to focus on how things are done *now* (as much as historical context is nice sometimes).

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  contents: read
+
+jobs:
+  build-sytest-images:
+    name: "Build sytest:${{ matrix.tag }}"
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+          - base_image: ubuntu:bionic
```
[the remark this answers is not in the document]
Comment (richvdh) [reply .github/workflows/docker.yml:20]: I generally prefer debian to ubuntu for this sort of thing - it has a longer support lifetime than the non-LTS ubuntus, and is bettter maintained than the LTS ones

suggest `debian:bullseye-slim`.

[the remark this answers is not in the document]
Comment (DMRobertson) [reply .github/workflows/docker.yml:20]: Can do. (I was just trying to replicate the existing behaviour as faithfully as I could.)

File: .github/workflows/docker.yml (hunk 2 above)
Comment (DMRobertson) [reply .github/workflows/docker.yml:38]: I could see the org-wide secret `DOCKER_HUB_USERNAME` is available to sytest, but I couldn't see one for the token. I was cribbing from [synapse's configuration](https://github.com/matrix-org/synapse/commit/c6509991f362cb559efbb97e1799776cd32a43d8#diff-3f5366f6d6df3ec1179e5efadc6f350bfa88eebf4e2da589b4d94ccb85ae5e94R34-R35)

File: docker/README.md (hunk 3 above)
Comment (DMRobertson) [reply docker/README.md:113]: It was useful for me in #1133 to be able to build these containers. (I'm not able to run sytest directly from my checkout, due to #1114). But I don't think that's a very persuasive argument for keeping `build.sh` around.

File: .github/workflows/docker.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply .github/workflows/docker.yml:20]: On second thoughts, I'll make that a second PR.

**Comment by reivilibre** [comment]
LGTM otherwise (though we should get ops to put the secret in before merge)

File: docker/README.md [hunk 5]
```diff
@@ -107,8 +107,8 @@ docker run --rm -it ... matrixdotorg/sytest-synapse:buster tests/20profile-event
 
 ## Building the containers
 
-The containers are built by executing `./build.sh`. You will then have to push
-them up to Docker Hub with `./push.sh`.
+These are automatically build and deployed on the `develop` branch by GitHub Actions.
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply docker/README.md:110]: One byte change to fix typo, thanks Olivier

[the remark this answers is not in the document]
Comment (DMRobertson) [reply docker/README.md:110]: This isn't a commit message

[the remark this answers is not in the document]
Comment (reivilibre) [reply docker/README.md]: /me was confused from e-mail of first one

**Review by reivilibre — APPROVED** [review]

**Pull request merged**